src/coro/coroToTexpr.ml

```diff
@@ -95,77 +96,82 @@ let handle_locals ctx b cls states tf_args forbidden_vars econtinuation =
 	(* Again, treat function arguments as the special case that they are *)
 	List.iter (fun (v, _) ->
 		if is_used_across_multiple_states v.v_id then begin
+			fst_state.cs_writes <- IntSet.add v.v_id fst_state.cs_writes;
+
 			let field = mk_field (Printf.sprintf "_hx_hoisted%i" v.v_id) v.v_type null_pos null_pos in
 
 			Hashtbl.replace fields v.v_id field;
+			Hashtbl.replace fst_state.cs_mapped_local v.v_id v;
 		end) tf_args;
 
 	List.iter (fun state ->
-		let is_not_declared_in_state id =
-			List.exists (fun v -> v.v_id == id) state.cs_declarations |> not in
+
+		let get_or_create_local_mapping v =
+			match Hashtbl.find_opt state.cs_mapped_local v.v_id with
+			| Some v -> v
+			| None ->
+				let new_v = alloc_var VGenerated (Printf.sprintf "_hx_restored%i" v.v_id) v.v_type v.v_pos in
+				Hashtbl.replace state.cs_mapped_local v.v_id new_v;
+				new_v
+		in
 
 		let rec mapper e =
 			match e.eexpr with
 			| TVar (v, eo) when is_used_across_multiple_states v.v_id ->
+				if Option.is_some eo then
+					state.cs_writes <- IntSet.add v.v_id state.cs_writes;
+
 				let field = mk_field (Printf.sprintf "_hx_hoisted%i" v.v_id) v.v_type v.v_pos v.v_pos in
 
 				Hashtbl.replace fields v.v_id field;
+				Hashtbl.replace state.cs_mapped_local v.v_id v;
 				
 				{ e with eexpr = TVar (v, Option.map mapper eo) }
-			| TLocal v when is_used_across_multiple_states v.v_id && is_not_declared_in_state v.v_id ->
+			| TBinop ((OpAssign | OpAssignOp _) as op, ({ eexpr = TLocal v; } as elhs), erhs) when is_used_across_multiple_states v.v_id ->
```

Comment:
I'm always paranoid about something like `(v) = rhs`, though I'm not sure if that's even legal syntax. That's generally an issue with this kind of deep-pattern matching, so I usually try to instead work with `Texpr.skip` to avoid some such cases.